### PR TITLE
chore(build): refactor broccoli ts files

### PR DESCRIPTION
**Pull request by alexeagle**

### CODE REVIEW

File: tools/broccoli/broccoli.d.ts [hunk 1]
```diff
@@ -0,0 +1,36 @@
+/// <reference path="../typings/es6-promise/es6-promise.d.ts" />
```
Comment (mprobst) [inline tools/broccoli/broccoli.d.ts]: I think there's value in keeping the things apart that are actually separate broccoli plugins and live in separate repos. If we ever wanted to migrate these definitions to DefinitelyTyped (which we should), then that'd make more sense, wouldn't it?

File: tools/broccoli/broccoli.d.ts [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+interface Tree {
+  read(readTree: (other: Tree) => Promise<string>);
```
Comment (mprobst) [inline tools/broccoli/broccoli.d.ts]: Isn't readTree a function of string? `readTree: (string) => Promise<string>`?

Can you points to docs on this?

File: tools/broccoli/broccoli.d.ts [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    write(readTree: (tree: Tree) => Promise<string>, destDir: string): Promise<void>;
+  }
+  export = Writer;
+}
+
+declare module "broccoli-caching-writer" {
+  class CachingWriter {
+    updateCache(srcDirs: string | string[], destDir: string);
+
+    constructor(inputTrees: Tree[], options: Object);
```
Comment (mprobst) [inline tools/broccoli/broccoli.d.ts]: constructor first?

File: tools/broccoli/traceur/index.ts [hunk 4]
```diff
@@ -12,7 +13,7 @@ var xtend = require('xtend');
 class TraceurFilter extends Writer {
   static RUNTIME_PATH = traceur.RUNTIME_PATH;
 
-  constructor(private inputTree, private destExtension: string,
+  constructor(private inputTree: Tree, private destExtension: string,
```
Comment (mprobst) [inline tools/broccoli/traceur/index.ts]: would it make sense to define a union type in broccoli.d.ts, `type Treeish = string | Tree;`?

File: tools/broccoli/broccoli.d.ts [hunk 5]
```diff
@@ -0,0 +1,36 @@
+/// <reference path="../typings/es6-promise/es6-promise.d.ts" />
+
+interface Tree {
```
Comment (mprobst) [inline tools/broccoli/broccoli.d.ts]: This should be in a namespace, shouldn't it?

File: tools/broccoli/broccoli.d.ts (hunk 1 above)
Comment (alexeagle) [inline tools/broccoli/broccoli.d.ts]: FYI, I'm going to drop this. I agree that we should mirror the structure they would have in definitely typed.
Igor is now adding a broccoli.d.ts for the node module `broccoli`

**Pull request closed without merging**